### DISCUSSION TITLE
Labeled parameters

### DISCUSSION THREAD

**Discussion by andrebaltieri** [dotnet/csharplang#6126] [Language Ideas]
Ok, this won't make a huge difference in C#, but maybe you like it :) 

There are some cases which a parameter name make all sense for those who are consuming the function, but not inside. There are situations where we want to use the same name from parameter inside the funcition but with another type and we can't because the variable name is already in use.

Can we support different names for a same parameter? One visible only outside and other visible only inside?
Something like this:

```csharp
// Approach 1
string DoSomething(string name text) 
{
  // In this case, we can see text inside this function
  return text.ToUpper();
}

// Here IDE/compiler will show parameter as name
var result = DoSomething(name: "John John"); 
```

This can also be done using annotations:
```csharp
// Approach 2
string DoSomething([Label("name")]string text) 
{
  // In this case, we can see text inside this function
  return text.ToUpper();
}

// Here IDE/compiler will show parameter as name
var result = DoSomething(name: "John John");  // This could work together with named parameters:
```

**Comment by HaloFour** [comment]
Why wouldn't you want the parameter name to be `name` in both cases?  I'm having trouble understanding where this would be useful.

**Reply by yaakov-h** [reply]
When used right, this feature could vastly improve the readability of code at the call-site when invoked with named arguments,. Here's a simple example from the Swift docs:

```swift
func greet(person: String, from hometown: String) -> String {
    return "Hello \(person)!  Glad you could visit from \(hometown)."
}
print(greet(person: "Bill", from: "Cupertino"))
// Prints "Hello Bill!  Glad you could visit from Cupertino."
```

In C# this would approximately be (assuming a similar syntax):

```csharp
string Greet(string person, string from hometown)
{
    return $"Hello {person}! Glad you could visit from ${hometown}.";
}
Console.WriteLine(Greet(person: "Bill", from: "Redmond"));
```

**Reply by FaustVX** [reply]
IMO, `Greet(person: "Bill", hometown: "Redmond")` is exactly as readable as `Greet(person: "Bill", from: "Redmond")`.

**Reply by HaloFour** [reply]
@yaakov-h 

Swift/Obj-C have a very different philosophy for method invocations, intending them to be read like a sentence.  That's why it makes more sense for the parameter to be given a label that is a preposition or another form of grammar that fits with a sentence structure.  But those languages also require the labels at the call site, and forbid reordering arguments, as otherwise the sentence would be gibberish.

In C# parameters are expected to be nouns, and can be reordered at will by the caller.  I find the parameter name `hometown` as readable in your example, if not moreso as it doesn't require the developer to infer what the parameter means based on the context of the grammar and surrounding parameter names.

**Reply by yaakov-h** [reply]
Yep, that's where this largely falls apart in C# - without that same callsite structure or being part of the method signature (i.e. allowing overloads that differ by label only), I don't see any advantage to this over creating a local variable immediately inside the function, i.e.:

```csharp
void Foo(string externalLabel)
{
    var internalName = externalLabel; // can probably use ref if its an expensive large struct
    // ...
}
```
